alot/__main__.py

```diff
@@ -13,115 +14,67 @@
-    ]
-    optFlags = [
-        ['omit_signature', '', 'do not add signature'],
-    ]
-
-    def parseArgs(self, *args):
-        SubcommandOptions.parseArgs(self, *args)
-        self.rest = ' '.join(args) or None
-
-
-class SearchOptions(SubcommandOptions):
-    accepted = ['oldest_first', 'newest_first', 'message_id', 'unsorted']
-
-    def colourint(val):
-        if val not in accepted:
-            raise ValueError("Unknown sort order")
-        return val
-    colourint.coerceDoc = "Must be one of " + str(accepted)
-    optParameters = [
-        ['sort', 'newest_first', None, 'Sort order'],
-    ]
-
-
-class Options(usage.Options):
-    optFlags = [["read-only", "r", 'open db in read only mode'], ]
-
-    def colourint(val):
-        val = int(val)
-        if val not in [1, 16, 256]:
-            raise ValueError("Not in range")
-        return val
-    colourint.coerceDoc = "Must be 1, 16 or 256"
-
-    def debuglogstring(val):
-        if val not in ['error', 'debug', 'info', 'warning']:
-            raise ValueError("Not in range")
-        return val
-    debuglogstring.coerceDoc = "Must be one of debug,info,warning or error"
-
-    optParameters = [
-        ['config', 'c', None, 'config file'],
-        ['notmuch-config', 'n', None, 'notmuch config'],
-        ['colour-mode', 'C', None, 'terminal colour mode', colourint],
-        ['mailindex-path', 'p', None, 'path to notmuch index'],
-        ['debug-level', 'd', 'info', 'debug log', debuglogstring],
-        ['logfile', 'l', '/dev/null', 'logfile'],
-    ]
-    search_help = "start in a search buffer using the querystring provided "\
-                  "as parameter. See the SEARCH SYNTAX section of notmuch(1)."
-
-    subCommands = [['search', None, SearchOptions, search_help],
-                   ['compose', None, ComposeOptions, "compose a new message"]]
-
-    def opt_version(self):
-        print alot.__version__
-        sys.exit(0)
-
 
 def main():
-    # interpret cml arguments
-    args = Options()
-    try:
-        args.parseOptions()  # When given no argument, parses sys.argv[1:]
-    except usage.UsageError as errortext:
-        print '%s' % errortext
-        print 'Try --help for usage details.'
-        sys.exit(1)
+    # set up the parser to parse the command line options.
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('-v', '--version', action='version',
+                        version='%(prog)s, {}'.format(alot.__version__))
+    parser.add_argument('-r', '--read-only', action='store_true',
+                        help='open db in read only mode')
+    parser.add_argument('-c', '--config', type=argparse.FileType('r'),
+                        help='config file')
+    parser.add_argument('-n', '--notmuch-config', type=argparse.FileType('r'),
+                        help='notmuch config')
+    parser.add_argument('-C', '--colour-mode', choices=(1, 16, 256), type=int,
```

Comment:
about default: what is the default here? I was assuming that it is dependent on the terminal type we detect and I hoped that urwid would do the right thing if we pass it None.

About being computer-sciency: I like that, don't apologize :) (but I did not get everything yet, see below)

About set: A set is a cool idea here. Will use it.

About the difference between list and tuples: I did not get that in total. I currently assume that python lists and tuples are essentially both linearly ordered aggregations/sequences/lists/you-give-it-a-mathematically-well-defined-name of arbitrary python objects, that allow for iteration (in for loops), random access with integer indices (in contrast to dicts which allow arbitrary indices), repeated access (contrast to generators, so they are stored and kept in memory and not generated on the fly), storing objects more than once in different positions (contrast to sets which allow each object only once). Possibly more? The difference being that lists are mutable (add/replace/remove elements) and tuples are not. So in order to relate this to you comment: how are lists more homogeneous than tuples? What kind of structure do tuples have that lists don't? Don't they both have an order? What does it mean that all elements of lists "quack"? (If this getting to long for a code-comment I am happy to move the discussion about lists vs. tuples to IRC.)